src/dfx/src/lib/nns/install_nns.rs

```diff
@@ -1,32 +1,661 @@
+        cmd.arg(account);
+    });
+    opts.sns_subnets.iter().for_each(|subnet| {
+        cmd.arg("--sns-subnet");
+        cmd.arg(subnet);
+    });
     cmd.stdout(std::process::Stdio::inherit());
     cmd.stderr(std::process::Stdio::inherit());
     let output = cmd
         .output()
         .with_context(|| format!("Error executing {:#?}", cmd))?;
 
     if !output.status.success() {
-        bail!("ic-nns-init call failed");
+        return Err(anyhow!("ic-nns-init call failed"));
+    }
+    Ok(())
+}
+
+/// Sets the exchange rate between ICP and cycles.
+///
+/// # Implementation
+/// This is done by proposal.  Just after startung a test server, ic-admin
+/// proposals with a test user pass immediately, as the small test neuron is
+/// the only neuron and has absolute majority.
+pub fn set_xdr_rate(rate: u64, nns_url: &Url) -> anyhow::Result<()> {
+    std::process::Command::new("ic-admin")
+        .arg("--nns-url")
+        .arg(nns_url.as_str())
+        .arg("propose-xdr-icp-conversion-rate")
+        .arg("--test-neuron-proposer")
+        .arg("--summary")
+        .arg(format!("Set the cycle exchange rate to {rate}."))
+        .arg("--xdr-permyriad-per-icp")
+        .arg(format!("{}", rate))
+        .stdin(process::Stdio::null())
+        .output()
+        .map_err(anyhow::Error::from)
+        .and_then(|output| {
+            if output.status.success() {
+                Ok(())
+            } else {
+                Err(anyhow!("Call to propose to set xdr rate failed"))
+            }
+        })
+}
+
+/// Sets the subnets the CMC is authorized to create canisters in.
+pub fn set_cmc_authorized_subnets(nns_url: &Url, subnet: &str) -> anyhow::Result<()> {
+    std::process::Command::new("ic-admin")
+        .arg("--nns-url")
+        .arg(nns_url.as_str())
+        .arg("propose-to-set-authorized-subnetworks")
+        .arg("--test-neuron-proposer")
+        .arg("--proposal-title")
+        .arg("Set Cycles Minting Canister Authorized Subnets")
+        .arg("--summary")
+        .arg(format!(
+            "Authorize the Cycles Minting Canister to create canisters in the subnet '{subnet}'."
+        ))
+        .arg("--subnets")
+        .arg(subnet)
+        .stdin(process::Stdio::null())
+        .output()
+        .map_err(anyhow::Error::from)
+        .and_then(|output| {
+            if output.status.success() {
+                Ok(())
+            } else {
+                Err(anyhow!("Call to propose to set xdr rate failed"))
+            }
+        })
+}
+
+/// Uploads wasms to the nns-sns-wasm canister.
+pub fn upload_nns_sns_wasms_canister_wasms(env: &dyn Environment) -> anyhow::Result<()> {
+    for SnsCanisterInstallation {
+        upload_name,
+        wasm_name,
+        ..
+    } in SNS_CANISTERS
+    {
+        // TODO: The sns binary needs to be bundled with dfx
+        let wasm_path = nns_wasm_dir(env).join(wasm_name);
+        let mut command = Command::new("sns");
```

Comment:
`sns` is available in the binary cache.  You can get it with env.get_cache().get_binary_command_path("sns")?;